Load Testcontainers parent first and start the containers with the Platform CL as the TCCL

gsmet opened the pull request:
> [!NOTE]
> This PR is part of a coordinated effort to address the OOM reported in https://github.com/quarkusio/quarkus/issues/42471:
> - https://github.com/quarkusio/quarkus/pull/42492
> - https://github.com/quarkusio/quarkus/pull/42525
> - https://github.com/quarkusio/quarkus/pull/42560

This is the last piece of the puzzle to make https://github.com/quarkusio/quarkus/issues/42471 actually pass.

TBH, I'm not exactly excited about adding testcontainers always but for now that's the only solution I could find.

There are two aspects to it:
- Start the containers with the TCCL set to the Platform CL to avoid having the TCCL of the threads being a `QuarkusClassLoader`
- Load Testcontainers base classes parent first: some are long lived so we need to avoid loading them with the `QuarkusClassLoader`

The problem with the latter is that we then need to make testcontainers a dependency of `quarkus-test-common` so that it appears in the test classpath.

There might be a better way to do it in the future with the work @holly-cummins has started but that's how it is for now (and I'm not entirely sure Holly's work would be enough to solve this as we would basically need to figure out that testcontainers is around and add it to a base class loader that would be shared across all the tests - it might be possible but I haven't looked at Holly's work yet).

The problem with Testcontainers is that there are some long lived threads that are kept around and our class loaders leak in two ways:
- by being the CL of some classes that are kept around
- by being the TCCL of some Testcontainers thread (Testcontainers itself, the Ryuk reaper thread and a ducttape thread)

Thus why this work is in two parts.

Class loader of a long lived class:

![Screenshot from 2024-08-16 17-00-37](https://github.com/user-attachments/assets/fa50ae71-4308-4b52-a8ba-06a00e2e10e3)


TCCL of thread:

![Screenshot from 2024-08-16 17-02-57](https://github.com/user-attachments/assets/7ba7f3ff-e3b8-462f-a5cf-01dc5ca781a8)

gsmet commented:
@aloubyansky this is what I was talking about.

gsmet commented on core/deployment/pom.xml:
```diff
@@ -18,6 +18,14 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-classloader-commons</artifactId>
         </dependency>
+        <!-- Uncomment this dependency to have Testcontainers loaded parent first -->
+        <!-- Make sure you build with -Denforcer.skip after that -->
+<!--
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>testcontainers</artifactId>
+        </dependency>
+-->
```
This is what somehow fixes the issue but I don't really want to do that.

gsmet commented on core/runtime/pom.xml:
```diff
@@ -257,6 +266,15 @@
+                          Testcontainers support: Testcontainers start a lot of threads that are long lived,
+                          leaking class loaders like crazy.
+                        -->
+                        <runnerParentFirstArtifact>org.testcontainers:testcontainers</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>com.github.docker-java:docker-java-api</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>com.github.docker-java:docker-java-transport-zerodep</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>org.rnorth.duct-tape:duct-tape</runnerParentFirstArtifact>
```
I'm not sure if this is actually needed... I tried things...

aloubyansky commented:
`quarkus-core-deployment` classes are a part of the bootstrap process, so they are on the classpath that launches the application, unlike the rest of the deployment dependencies. Actually, `quarkus-core-deployment` is also available for loading, along with other deployment dependencies, by the QuarkusClassLoader. It will depend on the classloader delegation whether that will happen.

gsmet committed:
Make Testcontainers parent first again

To avoid long lived threads being attached to child class loaders.
It was done at some point but was reverted because of the ducttape
dependency not being parent first.
I think it's worth another try by including this dependency.

gsmet committed:
Add testcontainers as a dependency of test-common

While we probably want to be a lot more clever about it, it's the
easiest way to have it in the test classpath and thus being allowed to
load it parent first.

gsmet committed:
Start Testcontainers containers with the Platform CL as TCCL

Testcontainers start several long lived threads (Ryuk reaper, ducttape),
and if the TCCL is one of the QuarkusClassLoader, we end up leaking the
class loaders.
Using the Platform CL makes sure we don't leak any QuarkusClassLoader
via these threads.

The observability containers have not been adjusted as they are a bit of
a weird beast and not worth having a look in the scope of this
experiment.

gsmet commented:
FWIW, while it fixes the problem in the OP's case and somehow works in some cases (a PostgreSQL container works fine), this PR comes with all sorts of issues.

It includes the following issues but there are others:

## MongoDB

```
2024-08-16T20:30:22.9762024Z Caused by: java.lang.IllegalAccessError: class org.testcontainers.containers.MongoDBContainer$MongoDBContainerDef cannot access its superclass org.testcontainers.containers.ContainerDef (org.testcontainers.containers.MongoDBContainer$MongoDBContainerDef is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @51508029; org.testcontainers.containers.ContainerDef is in unnamed module of loader 'app')
2024-08-16T20:30:22.9766481Z 	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
2024-08-16T20:30:22.9767668Z 	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
2024-08-16T20:30:22.9771695Z 	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
2024-08-16T20:30:22.9773679Z 	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:491)
2024-08-16T20:30:22.9775470Z 	at io.quarkus.mongodb.deployment.DevServicesMongoProcessor.lambda$startMongo$0(DevServicesMongoProcessor.java:184)
2024-08-16T20:30:22.9776915Z 	at java.base/java.util.Optional.orElseGet(Optional.java:364)
2024-08-16T20:30:22.9778302Z 	at io.quarkus.mongodb.deployment.DevServicesMongoProcessor.startMongo(DevServicesMongoProcessor.java:209)
2024-08-16T20:30:22.9780123Z 	at io.quarkus.mongodb.deployment.DevServicesMongoProcessor.startMongo(DevServicesMongoProcessor.java:113)
```

## IBM Db2

```
2024-08-16T18:21:48.1732198Z Caused by: java.lang.IllegalStateException: The image icr.io/db2_community/db2:11.5.9.0 requires you to accept a license agreement. Please place a file at the root of the classpath named container-license-acceptance.txt, e.g. at src/test/resources/container-license-acceptance.txt. This file should contain the line:
2024-08-16T18:21:48.1735318Z   icr.io/db2_community/db2:11.5.9.0
2024-08-16T18:21:48.1736670Z 	at org.testcontainers.utility.LicenseAcceptance.assertLicenseAccepted(LicenseAcceptance.java:30)
2024-08-16T18:21:48.1738487Z 	at org.testcontainers.containers.Db2Container.configure(Db2Container.java:83)
2024-08-16T18:21:48.1740496Z 	at io.quarkus.devservices.db2.deployment.DB2DevServicesProcessor$QuarkusDb2Container.configure(DB2DevServicesProcessor.java:98)
```

## Microsoft SQL Server

```
2024-08-16T18:22:21.6021940Z Caused by: java.lang.IllegalStateException: The image mcr.microsoft.com/mssql/server:2022-latest requires you to accept a license agreement. Please place a file at the root of the classpath named container-license-acceptance.txt, e.g. at src/test/resources/container-license-acceptance.txt. This file should contain the line:
2024-08-16T18:22:21.6025146Z   mcr.microsoft.com/mssql/server:2022-latest
2024-08-16T18:22:21.6026699Z 	at org.testcontainers.utility.LicenseAcceptance.assertLicenseAccepted(LicenseAcceptance.java:30)
2024-08-16T18:22:21.6028452Z 	at org.testcontainers.containers.MSSQLServerContainer.configure(MSSQLServerContainer.java:77)
2024-08-16T18:22:21.6030600Z 	at io.quarkus.devservices.mssql.deployment.MSSQLDevServicesProcessor$QuarkusMSSQLServerContainer.configure(MSSQLDevServicesProcessor.java:100)
```

gsmet commented:
I marked this PR ready for review to prepare for a talk (and get a CI report). DON'T MERGE IT.
